Add messages describing microgrid entities

tiyash-basu-frequenz opened the pull request:
This change introduces messages to describe microgrid entities such as sensors, components, electrical connections, and microgrids.

tiyash-basu-frequenz committed:
Add a message representing entity lifetimes

This message is a wrapper over the start and end timestamps of an
entity, e.g., component, connection, etc. It is used to indicate the
operational period of an entity.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add a message representing components

This commit adds a new message `ComponentCategoryMetadataVariant` which is
a union type for the metadata/description messages for different component
categories. Each component-category-specific metadata/description message
contains the fields required to minimally describe the component category
further, e.g., its type, or its rated max max current. These metadata
messages can be extended in the future.

In addition to the above, this commit also adds a message `Component`, that
describes components present in a microgrid. This message uses
`ComponentCategoryMetadataVariant` to further describe the component
metadata.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Update the packages of component categories

This commit updates the paths of protobuf files for component categories
to be more specific and to be more in line with the naming conventions of
the other packages.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add a message representing sensors

This commit adds a new message `Sensor`, which contains static information
about sensors installed in microgrids.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add message representing electrical connections between components

This commit adds a new message `ComponentConnection` to represent
electrical connections between two components installed in a microgrid.
The message is not about data flow but rather about the physical
electrical connections between components. Therefore, the IDs for the
source and destination components correspond to the actual setup within
the microgrid.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add a message representing delivery-areas

This commit adds the following definition:
`DeliveryArea`: A delivery area is a geographical or administrative
region, usually defined and maintained by a Transmission System
Operator (TSO), where electricity deliveries for a contract occur.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add a message describing microgrids

This commit adds the following definition:

`Microgrid`: A microgrid is a localized grouping of electricity
generation, energy storage, and loads that normally operates connected
to a traditional centralized grid. Each microgrid has a unique
identifier and is associated with an enterprise account. A key feature
is that it has a physical location and is situated in a delivery area.
Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz commented on proto/frequenz/api/common/v1/grid.proto:
```diff
@@ -0,0 +1,62 @@
+
+syntax = "proto3";
+
+package frequenz.api.common.v1.grid;
```
This package name somehow looks odd. This focuses mainly on market delivery areas (and delivery periods from the Assets API), but is called `grid`.

Also, as a part of cleaning up the package names (subject to review in this PR), I added a file `components/grid.proto` to represent grid connection points in a microgrid. Having two packages named `components.grid` and `grid` could be confusing.

I have the following suggestions:
1. Rename this package, e.g. to `trade_delivery` or something similar, OR
2. Place the delivery related messages in `microgrid.proto`, because they seem to be tighly related to the microgrid domain, OR
3. Rename the package `components.grid` to `components.grid_connection_point`.

Which one would you suggest? Or something else perhaps?

thomas-nicolai-frequenz replied:
> Rename this package, e.g. to trade_delivery or something similar, OR

Well these things are given/provided by the grid operator and not necessary just related to trading. The assets API is also not really about trading at all. These things are really related to the grid. They only apply if a microgrid is connected to a grid. I don't have a better idea right now tbh.

> Place the delivery related messages in microgrid.proto, because they seem to be tighly related to the microgrid domain

I'm not sure. `Counterparty` of example has nothing to do with the microgrid.

> Rename the package components.grid to components.grid_connection_point.

This feels misleading. These things are like the config parameters for the grid that a microgrid is connected to. Its like the metadata of the grid and how its being operated.

thomas-nicolai-frequenz replied:
Also if you look at this import path that looks totally legit to me. I mean its the delivery area of the grid, isn't it? It describes a certain part of the grid.

> frequenz.api.common.v1.grid.DeliveryArea

tiyash-basu-frequenz replied:
Okay, so we leave it as is?

thomas-nicolai-frequenz replied:
Well I don't have a better idea right now.

tiyash-basu-frequenz replied:
Okay, then we can keep it as for now. We can revisit later if needed.

tiyash-basu-frequenz commented on proto/frequenz/api/common/v1/components.proto:
```diff
@@ -75,46 +82,100 @@ enum ComponentCategory {
+message ComponentConnection {
+  // Unique identifier of the component where the connection originates. This is
+  // aligned with the direction of current flow away from the grid connection
+  // point, or in case of islands, away from the islanding point.
```
I extended the original documentation slightly, in order to accommodate islands. If this looks good, I will add this to other places `ComponentConnection`'s documentation.

tiyash-basu-frequenz replied:
Added this elsewhere. Adjusted the phrases slightly to go with the flow of the original documentation.

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/components/grid.proto:
```diff
@@ -0,0 +1,26 @@
+package frequenz.api.common.v1.components.grid;
+
+// A representation of a grid connection point.
+message Grid {
```
How about renaming this to `GridConnection` or `GridFuse` instead of just Grid? Grid isn't really a "component".

tiyash-basu-frequenz replied:
ChatGPT suggests that the most widely known name for this point is `Point of Common Coupling (PCC)`. However, I prefer the suggestion `Grid Interconnection Point`. What do you suggest?

![image](https://github.com/frequenz-floss/frequenz-api-common/assets/63402123/bacc088f-1331-4699-b56c-e22a89155eca)

tiyash-basu-frequenz replied:
Or just `GridConnectionPoint` maybe?

thomas-nicolai-frequenz replied:
https://en.wikipedia.org/wiki/Microgrid#Point_of_common_coupling_(PCC)

It looks like ChatGPT is right if you look at Wikipedia. Of course GridConnectionPoint is a more general term describing the same concept without using the more sophisticated terminology. Which way we ever go we can of course also use the documentation to clarify. Why don't you put out a vote in Rnd or something?

tiyash-basu-frequenz replied:
As discussed in slack, changed it to `GridConnectiopnPoint`. The idea is to be more approachable to users that do not have an EE background.

thomas-nicolai-frequenz replied:
Yes, however don't forget to mention PCC in the documentation and briefly explain why we choose grid connection point (GCP) over PCC.

thomas-nicolai-frequenz commented on proto/frequenz/api/common/v1/lifetime.proto:
```diff
@@ -0,0 +1,28 @@
+
+syntax = "proto3";
+
+package frequenz.api.common.v1.lifetime;
```
Why wouldn't we put that into microgrid.proto? I don't feel this should deserve its own file.

> Lifetime captures the timestamps indicating the active operational period of a microgrid asset

tiyash-basu-frequenz replied:
We discussed it to be in `lifetime.proto` since this is being used in several places, including `Component` and `ComponentConnection`.

thomas-nicolai-frequenz replied:
Sure also sensors but these are all part of the microgrid. This has no application beyond microgrids. Hence the question. The most outer scope of this is the microgrid.

tiyash-basu-frequenz replied:
That actually raises a point: should we package microgrid-specific entities in a directory named `microgrid`? That would address the issue with scoping that you brought up.

Otherwise, just putting `Lifetime` in `microgrid.proto`, but leaving the rest out feels arbitrary ..

thomas-nicolai-frequenz replied:
> should we package microgrid-specific entities in a directory named microgrid

we can of course but in this case we should apply the same to all the others. it also raises the question what the structure in these subfolders look like My vote would still got to put it into the microgrid.proto though. At least for now until we have established a clear structure for everything. I don't feel good moving on some parts while not moving on others.

tiyash-basu-frequenz replied:
As discussed in-person, I updated the file structure for microgrid related entities and grid/delivery_area.

Ready for a look.

I will touch the rest (location, metrics, pagination) in a different PR.

tiyash-basu-frequenz replied:
This is the structure we have for grid, and microgrid related files:

```
grid/
  delivery_area.proto (package: frequenz.api.common.v1.grid)
microgrid/
  components/
    battery.proto (package: frequenz.api.common.v1.microgrid.components)
    ev_charger.proto (package: frequenz.api.common.v1.microgrid.components)
    fuse.proto (package: frequenz.api.common.v1.microgrid.components)
    grid.proto (package: frequenz.api.common.v1.microgrid.components)
    inverter.proto (package: frequenz.api.common.v1.microgrid.components)
    transformer.proto (package: frequenz.api.common.v1.microgrid.components)
    components.proto (package: frequenz.api.common.v1.microgrid.components)
  sensors/
    sensors.proto (package: frequenz.api.common.v1.microgrid.sensors)
  lifetime.proto (package: frequenz.api.common.v1.microgrid)
```

So, the package of each protobuf file is also its path in the file structure.

Again, `locations.proto`, `metrics.proto`, and `pagination.proto` will be updated in a different PR.

thomas-nicolai-frequenz replied:
I think thats reasonable. I'm still not happy with the lifetime.proto but I don't have a better idea, so lets go with that!

tiyash-basu-frequenz commented on proto/frequenz/api/common/v1/components.proto:
```diff
@@ -75,46 +82,102 @@ enum ComponentCategory {
+// Microgrid electrical component details.
+message Component {
+  // The component ID.
+  uint64 id = 1;
```
Forgot to mention earlier: in the original specs, the `id` field has the type `ComponentId`. Since the definition of a component ID was missing, I added the type `uint64`.

tiyash-basu-frequenz committed:
Update proto file paths and packages related to microgrids

The following package renames have been done:

- `frequenz.api.common.v1.components` ->
   `frequenz.api.common.v1.microgrid.components`

- `frequenz.api.common.v1.lifetime` ->
   `frequenz.api.common.v1.microgrid.lifetime`

- `frequenz.api.common.v1.sensors` ->
   `frequenz.api.common.v1.microgrid.sensors`

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Update proto file paths for grid/delivery_area

The file `grid.proto` contained a `DeliveryArea` message. This file has
now been renamed to `grid/delivery_area.proto`.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>
